api/python/t4/formats.py

```diff
@@ -0,0 +1,1007 @@
+so variances in format between physical keys cannot be tolerated, unless there
+is also a change to what data the logical key references.
+
+A FormatHandler has, at bare minimum:
+    * a name specific to the format used (NOT to the format handler used)
+      * I.e., two FormatHandler objects that both handle JSON should both be
+        named 'json'
+    * a serializer
+    * a deserializer
+
+Aside from that, a FormatHandler *should* have:
+    * a list of filename extensions it can (theoretically) handle
+    * a list of object types it can handle
+
+Format objects can be registered directly by calling "f.register()".
+
+
+# Format metadata
+
+In an object's metadata, the format should only touch the 'format' key,
+and possibly the 'target' key.
+
+Format metadata has the following form:
+
+```
+{
+  # 'name':
+  #     a unique format name, like csv, json, parquet, numpy, etc
+  'name': 'csv',
+
+  # 'opts', I.e. Format Options:
+  #     Options to help in serialization/deserialization.
+  #     * opts are needed when a format is leaky/ill-defined, as with CSV.
+  #     * opts should not exactly match underlying serializer/deserializer
+  #       args unless known to be safe or analyzed at runtime for safety.
+  #     * opts should be kept as platform-independent as possible.
+  #     * option names must be present in <format_handler>.opts, or a warning
+  #       will be shown, and the option will be ignored.
+  'opts': {<option name>: <option value>}
+}
+
+"""
+
+
+# Python imports
+from abc import ABC, abstractmethod
+from collections import Mapping
+import copy
+import csv
+import io
+import json
+import sys
+import warnings
+
+# Third Party imports
+from six import text_type
+
+# Project imports
+from .util import QuiltException
+
+
+# Constants
+NOT_SET = type('NOT_SET', (object,), {'__doc__':
+    """A unique indicator of disuse when `None` is a valid value"""
+    })()
+
+
+# Code
+class FormatRegistry:
+    """A collection for organizing `FormatHandler` objects.
+
+    This class organizes `FormatHandler` objects for querying and general use.
+    It provides methods for querying by format name, metadata dict, handled
+    extensions, or handled object types.  This list may expand in the future,
+    so see the actual class methods.
+    """
+    registered_handlers = list()
+
+    # latest adds are last, and come first in lookups by type via `for_obj`.
+    def __init__(self):
+        raise TypeError("The {!r} class is organizational, and cannot be instantiated."
+                        .format(type(self).__name__))
+
+    @classmethod
+    def register(cls, handler):
+        """Register a FormatHandler instance"""
+        handlers = cls.registered_handlers
+
+        # no duplicates, just reprioritize.
+        if handler in handlers:
+            handlers.pop(handlers.index(handler))
+
+        handlers.insert(0, handler)
+
+    @classmethod
+    def search(cls, obj_type=None, meta=None, ext=None, single=True, handlers=None):
```

Comment:
Please remove `single` from all the functions - it's weird to have a return type that depends on arguments.

If the list return type gets annoying, we can add helper functions that return a single item.

Reply:
Done.